larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,470 @@
+    const ClusterList *pShowerClusters(nullptr);
+    PandoraContentApi::GetList(*this, "ShowerClusters3D", pShowerClusters);
+    if(!pShowerClusters) return false;
+
+    if(pShowerClusters->end() == std::find(pShowerClusters->begin(), pShowerClusters->end(), clusterList3D.front())) return false;
+    if(this->GetFigureOfMerit(caloHitList3D)<m_fomThresholdForReclustering) return false; 	
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float ThreeDReclusteringAlgorithm::GetCheatedFigureOfMerit(CaloHitList mergedClusterCaloHitList3D)
+{
+    std::map<int,int> mainMcParticleMap;
+    for (const CaloHit *const pCaloHit : mergedClusterCaloHitList3D)
+    {
+        const CaloHit *const pParentCaloHit = static_cast<const CaloHit *>(pCaloHit->GetParentAddress());
+        int mainMcParticleIndex = this->GetMainMcParticleIndex(pParentCaloHit);
+        std::map<int, int>::iterator it = mainMcParticleMap.find(mainMcParticleIndex);
+ 
+        if (it != mainMcParticleMap.end()) {
+            it->second++;
+        }
+        else {
+            mainMcParticleMap.insert(std::make_pair(mainMcParticleIndex, 1));
+        }
+    }
+    const MCParticleList *pMCParticleList(nullptr);
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*this, m_mcParticleListName, pMCParticleList));
+    MCParticleVector mcParticleVector(pMCParticleList->begin(),pMCParticleList->end());
+    std::sort(mcParticleVector.begin(), mcParticleVector.end(), LArMCParticleHelper::SortByMomentum);
+    
+    auto maxSharedHits = std::max_element(mainMcParticleMap.begin(), mainMcParticleMap.end(), [](const auto &x, const auto &y) {return x.second < y.second;});
+    float mainMcParticleFraction = (float)maxSharedHits->second/mergedClusterCaloHitList3D.size();
+    return (1-mainMcParticleFraction);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+int ThreeDReclusteringAlgorithm::GetMainMcParticleIndex(const pandora::CaloHit *const pCaloHit)
+{
+    const MCParticleList *pMCParticleList(nullptr);
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*this, m_mcParticleListName, pMCParticleList));
+    MCParticleVector mcParticleVector(pMCParticleList->begin(),pMCParticleList->end());
+    std::sort(mcParticleVector.begin(), mcParticleVector.end(), LArMCParticleHelper::SortByMomentum);
+    int iMcPart(0); 
+    for (const auto &weightMapEntry : pCaloHit->GetMCParticleWeightMap()) 
+    { 
+      if(weightMapEntry.second>0.5) 
+      { 
+        iMcPart=0;  
+        for(const MCParticle *const pMCParticle: mcParticleVector) 
+        { 
+            if(pMCParticle==weightMapEntry.first) { break;} 
+            iMcPart++; 
+        }
+      } 
+    }
+    return iMcPart;
+}
```

Comment:
Since this is used in a few places, should this (or maybe an equivalent does?) live in an MC / CaloHit helper file?

Reply:
Where do we want this to live? A CaloHitHelper seems like the most appropriate place, although this function only makes sense for sim. I don't think a CaloHitHelper exists yet. I can make one, unless there is a different preference.

Reply:
To add to this, there are already functions for this for extracting the actual MC particle, e.g. MCParticleHelper::GetMainMCParticle in the SDK. So if you are happy working directly with the MC particle pointers instead of an integer index, this is an option (or construct a map between MC pointers and the desired integer index)